stellar/utils.ts

```diff
@@ -186,34 +192,54 @@ async function sendTransaction(tx, server, action, options: Options = {}) {
     }
 }
 
-async function broadcast(operation, wallet, chain, action, options: Options, simulateTransaction = false) {
+async function broadcast(
+    operation: xdr.Operation,
+    wallet: Keypair,
+    chain,
+    action: string,
+    options: Options,
+    simulateTransaction?: boolean,
+): Promise<xdr.ScVal | rpc.Api.SimulateTransactionSuccessResponse> {
+    /* Allows user to require or disallow simulation. Will always attempt if not specified. */
+    if (simulateTransaction === undefined) simulateTransaction = true;
+
     const server = new rpc.Server(chain.rpc, { allowHttp: chain.networkType === 'local' });
+    const tx = await buildTransaction(operation, server, wallet, chain.networkType, options);
+    const response = await server.simulateTransaction(tx);
+
+    if (simulateTransaction && isReadOnly(response)) {
+        if (rpc.Api.isSimulationError(response)) {
+            printError('Failed to simulate transaction');
+            throw new Error(response.error);
+        }
+
+        printInfo('Successfully simulated tx', `action: ${action}, networkType: ${chain.networkType}, chainName: ${chain.name}`);
+        return response;
+    }
 
     if (options && options.nativePayment) {
-        const tx = await buildTransaction(operation, server, wallet, chain.networkType, options);
         tx.sign(wallet);
         return sendTransaction(tx, server, action, options);
     }
+
     if (options && options.estimateCost) {
-        const tx = await buildTransaction(operation, server, wallet, chain.networkType, options);
         const resourceCost = await estimateCost(tx, server);
         printInfo('Gas cost', JSON.stringify(resourceCost, null, 2));
         return;
     }
 
-    if (simulateTransaction) {
-        const tx = await buildTransaction(operation, server, wallet, chain.networkType, options);
-        try {
-            const response = await server.simulateTransaction(tx);
-            printInfo('successfully simulated tx', `action: ${action}, networkType: ${chain.networkType}, chainName: ${chain.name}`);
-            return response;
-        } catch (error) {
-            throw new Error(error);
-        }
-    }
+    const preparedTx = await prepareTransaction(operation, server, wallet, chain.networkType, options);
+    return sendTransaction(preparedTx, server, action, options);
+}
+
+function isReadOnly(response: rpc.Api.SimulateTransactionResponse): boolean {
+    if (!rpc.Api.isSimulationSuccess(response)) return false;
+    if (response.transactionData.getReadWrite().length > 0) return false;
+    if (response.result && response.result.auth.length > 0) return false;
+    if (response.stateChanges && response.stateChanges.length > 0) return false;
```

Comment:
May be try to optimize it something like this?

```suggestion
    if (
        !rpc.Api.isSimulationSuccess(response) ||
        response.transactionData.getReadWrite().length > 0 ||
        response.result?.auth?.length > 0 ||
        response.stateChanges?.length > 0
    ) {
        return false;
    }
```